syftlib/src/main/java/org/openmined/kotlinsyft/WebRTC.kt

```diff
@@ -0,0 +1,304 @@
+import org.webrtc.PeerConnection.*
+import java.nio.ByteBuffer
+
+
+typealias SDP_type = SessionDescription.Type
+
+private const val TAG = "WebRTCClient"
+
+open class WebRTCClient(
+    peerOptions: PeerConnectionFactory.Options,
+    private val peerConfig: RTCConfiguration, private val socket: Socket
+) {
+
+    private val peerConnectionFactory = PeerConnectionFactory.builder()
+        .setOptions(peerOptions)
+        .createPeerConnectionFactory()
+
+    var peers = HashMap<String, Peer>()
+    private lateinit var workerId: String
+    private lateinit var scopeId: String
+
+    fun start(workerId: String, scopeId: String) {
+        Log.d(TAG, "Joining room $scopeId")
+
+        this.workerId = workerId
+        this.scopeId = scopeId
+        //TODO send as json {workerID,scopeId}
+        socket.send(WEBRTC_JOIN_ROOM, "{$workerId,$scopeId}")
+    }
+
+    private fun createConnection(worker_id: String) {
+        Log.d(TAG, "Creating Connection as answer")
+        val pcObserver = PeerConnectionObserver(worker_id, SDP_type.ANSWER)
+        val pc = peerConnectionFactory.createPeerConnection(peerConfig, pcObserver)
+        peers[worker_id] = Peer(pc, null, pcObserver, SDPObserver(worker_id, SDP_type.ANSWER))
+    }
+
+    fun stop() {
+        Log.d(TAG, "WebRTC disconnecting from peers")
+        peers.forEach { (worker_id, peer) ->
+            if (peer.channel != null)
+                removePeer(worker_id)
+        }
+    }
+
+    private fun removePeer(worker_id: String) {
+        if (!peers.keys.contains(worker_id)) return
+
+        Log.d(TAG, "`Closing connection to $worker_id")
+        try {
+            peers[worker_id]!!.channel?.close()
+            //uncomment this if we need to close the connection as well
+            //peers[workerId]!!.connection.close()
+        } catch (e: Exception) {
+            Log.e(TAG, "error removing peer $worker_id", e)
+        }
+        peers.remove(worker_id)
+    }
+
+    // Given a message, this function allows you to "broadcast" a message to all peers
+    // Alternatively, you may send a targeted message to one specific peer (specified by the "to" param)
+    fun sendMessage(message: String, to: String?) {
+        Log.d(TAG, "sending message $message")
+
+        if (to != null && !to.contentEquals(workerId) && peers[to]?.channel != null)
+            send(peers[to]!!.channel!!, message)
+        else
+            peers.filterValues { it.channel != null }
+                .forEach { (_, peer) -> send(peer.channel!!, message) }
+    }
+
+    private fun send(channel: DataChannel, msg: String) {
+        try {
+            channel.send(DataChannel.Buffer(ByteBuffer.wrap(msg.toByteArray()), false))
+        } catch (e: Exception) {
+            Log.e(TAG, "error sending message", e)
+        }
+    }
+
+    private fun sendInternalMessage(type: String, message: String?, to: String) {
+        //TODO implement this
+    }
+
+    fun receiveNewPeer(worker_id: String) {
```

Comment:
All `worker_id` should be `workerId`

Reply:
This would almost always be different from global workerId. Should we still override variable names?

Reply:
Access to the member will be qualified with `this`. If in doubt, call it `newWorkerId`.

Reply:
ccol